Base/Python/slicer/parameterNodeWrapper/guiConnectors.py

```diff
@@ -0,0 +1,709 @@
+
+    def _connect(self):
+        self._widget.directoryChanged.connect(self.changed)
+
+    def _disconnect(self):
+        self._widget.directoryChanged.disconnect(self.changed)
+
+    def widget(self) -> ctk.ctkDirectoryButton:
+        return self._widget
+
+    def read(self):
+        return self._type(self._widget.directory)
+
+    def write(self, value) -> None:
+        self._widget.directory = str(value)
+
+
+@parameterNodeGuiConnector
+class qMRMLNodeComboBoxToNodeConnector(GuiConnector):
+    @staticmethod
+    def canRepresent(widget, datatype) -> bool:
+        return type(widget) == slicer.qMRMLNodeComboBox and isNodeOrUnionOfNodes(datatype)
+
+    @staticmethod
+    def create(widget, datatype):
+        if qMRMLNodeComboBoxToNodeConnector.canRepresent(widget, datatype):
+            return qMRMLNodeComboBoxToNodeConnector(widget, datatype)
+        return None
+
+    def __init__(self, widget: slicer.qMRMLNodeComboBox, datatype):
+        self._widget: slicer.qMRMLNodeComboBox = widget
+        self._widget.nodeTypes = getNodeTypes(datatype)
+
+    def _connect(self):
+        self._widget.currentNodeChanged.connect(self.changed)
+
+    def _disconnect(self):
+        self._widget.currentNodeChanged.disconnect(self.changed)
+
+    def widget(self) -> slicer.qMRMLNodeComboBox:
+        return self._widget
+
+    def read(self):
+        return self._widget.currentNode()
+
+    def write(self, value) -> None:
+        self._widget.setCurrentNode(value)
+
+
+@parameterNodeGuiConnector
+class qMRMLSubjectHierarchyTreeViewToNodeConnector(GuiConnector):
+    @staticmethod
+    def canRepresent(widget, datatype) -> bool:
+        return type(widget) == qMRMLSubjectHierarchyTreeView and isNodeOrUnionOfNodes(datatype)
+
+    @staticmethod
+    def create(widget, datatype):
+        if qMRMLSubjectHierarchyTreeViewToNodeConnector.canRepresent(widget, datatype):
+            return qMRMLSubjectHierarchyTreeViewToNodeConnector(widget, datatype)
+        return None
+
+    def __init__(self, widget: qMRMLSubjectHierarchyTreeView, datatype):
+        self._widget: qMRMLSubjectHierarchyTreeView = widget
+        self._widget.nodeTypes = getNodeTypes(datatype)
+
+    def _connect(self):
+        self._widget.currentItemsChanged.connect(self.changed)
+
+    def _disconnect(self):
+        self._widget.currentItemsChanged.disconnect(self.changed)
+
+    def widget(self) -> qMRMLSubjectHierarchyTreeView:
+        return self._widget
+
+    def read(self):
+        itemId = self._widget.currentItem()
+        shNode = self._widget.subjectHierarchyNode()
+        if itemId == shNode.GetInvalidItemID():
+            return None
+        else:
+            return shNode.GetItemDataNode(itemId)
+
+    def write(self, value) -> None:
+        if value is not None:
+            self._widget.setCurrentNode(value)
+        else:
+            self._widget.clearSelection()
+
+
+SlicerPackParameterNamePropertyName = "SlicerPackParameterName"
```

Comment:
Per other comment, this is essentially part of the WidgetChildrenToParameterPackConnector interface, so it was placed near that.

Reply:
This may be a point in favor of breaking this into a package, and moving implementations to their own module. This would allow the parameter pack connector to have its globals at the top of the module in good style, and avoid polluting the large file with its helpers.

Reply:
I can see pulling `WidgetChildrenToParameterPackConnector` into its own class because it is more complex than the others. Are you recommending something like a "one file per connector" setup in general?

Reply:
I think one file per connector is too granular. Perhaps one file per data type; eg group all the number connectors into one file, the string/path connectors etc?

Reply:
The `QComboBoxToStringableConnector` breaks that because it handles multiple data types. I agree breaking the file up could help. There are just a lot of exceptions that make coming up with a rule difficult.